[#6404] feat (gvfs-fuse): Add Readme document of gvfs-fuse (#6405)
apache/gravitino: clients/filesystem-fuse/README.md

```

## Usage Guide

Navigate to the `clients/filesystem-fuse` directory, build the executable binary and use the following commands to view
help information or to perform mounting operations:

```shell
# Display help information for gvfs-fuse
target/debug/gvfs-fuse --help

# Display help for the mount command
target/debug/gvfs-fuse mount --help

# Display help for the umount command
target/debug/gvfs-fuse umount --help

#create mount directory
mkdir -p target/gvfs

# Execute the mount command in the foreground
# mount the fileset uri to the local directory target/gvfs. You need to start the gravitino server and create the fileset first
target/debug/gvfs-fuse mount target/gvfs gvfs://fileset/test/c1/s1/fileset1 -c conf/gvfs_fuse.toml -f

# Execute the mount command in the background
# mount the fileset uri to the local directory target/gvfs
target/debug/gvfs-fuse mount target/gvfs gvfs://fileset/test/c1/s1/fileset1 -c conf/gvfs_fuse.toml

# Execute the umount command
# unmount the fileset from the local directory target/gvfs
target/debug/gvfs-fuse umount target/gvfs
```

The `conf/gvfs_fuse.toml` file is a configuration file that contains the following information:

> ```toml

Reviewer: could you add some comments about how to use memory filesystem for test?

Author: Add description after the example configuration

<!--
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->

# Gvfs-fuse

Gvfs-fuse is the Fuse client for Gravitino fileset. It allows users to mount Gravitino filesets to their local file system via Fuse, enabling access to Gravitino fileset files as if they were part of the local file system.

## Features

- Supports mounting S3 filesets.
- Supports basic file read and write operations.
- Supports directory operations

> - Allows file attribute manipulation.

Reviewer 2: None of the above items are full sentences, so the ending periods could be all removed.

Author: fixed

## Environment setup and build instructions

### Prerequisites

> You need to install the following software before you can build gvfs-fuse

Reviewer 2: ```suggestion
You need to install the following software before you can build gvfs-fuse:
```

- [fuse3 and libfuse3-dev](https://www.kernel.org/doc/html/next/filesystems/fuse.html) for Linux systems.
- [macFUSE](https://macfuse.github.io) for macOS systems.
- [Rust](https://www.rust-lang.org) environment for compiling.

### Build process

Navigate to the `client/filesystem-fuse` directory of the Gravitino project and execute the following commands:

```shell
# Code verification
make check

# Code formatting
make fmt

# Building Gvfs-fuse
make
```

> After building, the executable `gvfs-fuse` binary will be located in the target/debug directory.

Reviewer 2: ```suggestion
After building, the `gvfs-fuse` executable will be located in the `target/debug` directory.
```

### Testing

Run the following commands to execute tests:

```
# Run tests that do not depend on the S3 or Gravitino environment
make test

# Run tests that depend on the S3 environment
make test-s3

# Run integration tests that depend on the Gravitino fileset environment
make test-fuse-it

> ```

Reviewer 2: Are all the above tests necessary for all scenarios?

Author: No, you can run them by your chose.

Reviewer 2: Then please clarify the scenarios each of the test apply.

## Usage Guide

> Navigate to the `clients/filesystem-fuse` directory, build the executable binary and use the following commands to view

Reviewer 2: ```suggestion
Navigate to the `clients/filesystem-fuse` directory, run the following commands to view
```

#create mount directory
mkdir -p target/gvfs

# Execute the mount command in the foreground
# mount the fileset uri to the local directory target/gvfs. You need to start the gravitino server and create the fileset first
target/debug/gvfs-fuse mount target/gvfs gvfs://fileset/test/c1/s1/fileset1 -c conf/gvfs_fuse.toml -f

# Execute the mount command in the background
# mount the fileset uri to the local directory target/gvfs
target/debug/gvfs-fuse mount target/gvfs gvfs://fileset/test/c1/s1/fileset1 -c conf/gvfs_fuse.toml

# Execute the umount command
# unmount the fileset from the local directory target/gvfs
target/debug/gvfs-fuse umount target/gvfs
```

The `conf/gvfs_fuse.toml` file is a configuration file that contains the following information:

```toml
# fuse settings
[fuse]
file_mask= 0o600
dir_mask= 0o700
fs_type = "gvfs"

# filesystem settings
[filesystem]
block_size = 8192

# Gravitino settings
[gravitino]
uri = "http://localhost:8090"
metalake = "test"

# extend settings
[extend_config]
s3-access_key_id = "XXX_access_key"
s3-secret_access_key = "XXX_secret_key"
s3-region = "XXX_region"
s3-bucket = "XXX_bucket"
s3-endpoint = "XXX_endpoint"
```

> The command starts a FUSE program with the given configuration, using a `MemoryFilesystem` for testing.

Reviewer 2: Which command?

Author: fix

You can access the filesystem in the `target/gvfs` directory.

> More configration file examples can be found in the `test/conf` directory.

Author: fix

<!--
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->

# Gvfs-fuse

Gvfs-fuse is the Fuse client for Gravitino fileset. It allows users to mount Gravitino filesets to their local file system via Fuse, enabling access to Gravitino fileset files as if they were part of the local file system.

> ## Features

Reviewer: please add empty line before and after headers

Author: fixed

Reviewer: please correct all documents

- Supports mounting S3 filesets.
- Enables basic file read and write operations.
- Allows file attribute manipulation.

> - Facilitates directory operations within filesets.

Reviewer: change to `Supports directory operations`?

Author: fixed

## Environment Setup and Build Instructions
### Prerequisites

> - `libfuse3-dev` and `fuse3` for Linux systems.

Reviewer: could you provide more context about `macFUSE` `libfuse3-dev` and `fuse3` to let users use it easily.

Author: Add links for references

```shell
# Code verification
make check

# Code formatting
make fmt

# Building Gvfs-fuse
make
```

After building, the executable gvfs-fuse will be located in the target/debug directory.

### Testing

Run the following commands to execute tests:

```
# Run tests that do not depend on the S3 or Gravitino environment
make test

# Run tests that depend on the S3 environment
make test-s3

# Run integration tests that depend on the Gravitino fileset environment
make test-fuse-it
```

## Usage Guide

Navigate to the directory containing the executable and use the following commands to view help information
or to perform mounting operations:

```shell
# Display help information for gvfs-fuse
./gvfs-fuse --help

# Display help for the mount command
./gvfs-fuse mount --help

# Display help for the umount command
./gvfs-fuse umount --help

# Execute the mount command in the foreground
# mount the fileset uri to the local directory /mnt/gvfs
./gvfs-fuse mount /mnt/gvfs gvfs://fileset/test/c1/s1/fileset1 -c /etc/gvfs_fuse.toml -f

# Execute the mount command in the background
# mount the fileset uri to the local directory /mnt/gvfs

> ./gvfs-fuse mount /mnt/gvfs gvfs://fileset/test/c1/s1/fileset1 -c /etc/gvfs_fuse.toml

Reviewer: missing `unmount`?

Author: fixed

<!--
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->

# Gvfs-fuse

Gvfs-fuse is the Fuse client for Gravitino fileset. It allows users to mount Gravitino filesets to their local file system via Fuse, enabling access to Gravitino fileset files as if they were part of the local file system.

## Features

- Supports mounting S3 filesets.
- Supports basic file read and write operations.
- Supports directory operations
- Allows file attribute manipulation.

> ## Environment Setup and Build Instructions

Reviewer 3: Environment Setup and Build Instructions -> Environment setup and build instructions

### Prerequisites

> - [fuse3 and libfuse3-dev](https://www.kernel.org/doc/html/next/filesystems/fuse.html) for Linux systems.

Reviewer 3: Can these three software be installed on the machine directly? I suggest you add more sentence like 'you need to install the following software before you can use gvfs-fuse' or something similar.

- [macFUSE](https://macfuse.github.io) for macOS systems.
- [Rust](https://www.rust-lang.org) environment for compiling.

> ### Build Process

Reviewer 3: Build Process

Only the first letter of the first word needs to be capitalized unless it's a proper noun. ditto

Navigate to the `client/filesystem-fuse` directory of the Gravitino project and execute the following commands:

```shell
# Code verification
make check

# Code formatting
make fmt

# Building Gvfs-fuse
make
```

> After building, the executable gvfs-fuse will be located in the target/debug directory.

Reviewer 3: After building the source, the executable `gvfs-fuse` binary...

### Testing

Run the following commands to execute tests:

```
# Run tests that do not depend on the S3 or Gravitino environment
make test

# Run tests that depend on the S3 environment
make test-s3

# Run integration tests that depend on the Gravitino fileset environment
make test-fuse-it
```

## Usage Guide

> Navigate to the directory containing the executable and use the following commands to view help information

Reviewer 3: ... the executable .. ->  ...to executable binary...

or to perform mounting operations:

```shell
# Display help information for gvfs-fuse
./gvfs-fuse --help

# Display help for the mount command
./gvfs-fuse mount --help

# Display help for the umount command
./gvfs-fuse umount --help

# Execute the mount command in the foreground
# mount the fileset uri to the local directory /mnt/gvfs

> ./gvfs-fuse mount /mnt/gvfs gvfs://fileset/test/c1/s1/fileset1 -c /etc/gvfs_fuse.toml -f

Reviewer: is there any other place to describe how to config `gvfs_fuse.toml`?

Author: No, this document is primarily intended to guide developers on how to build, test, and run the application.

Reviewer: without `gvfs_fuse.toml`, how should developers run gvfs?

Author: Add some descriptions and examples
